pkg/objstore/azure/azure.go

```diff
@@ -0,0 +1,318 @@
+
+	for _, blob := range list.Segment.BlobItems {
+		listNames = append(listNames, blob.Name)
+	}
+
+	for _, blobPrefix := range list.Segment.BlobPrefixes {
+		listNames = append(listNames, blobPrefix.Name)
+	}
+
+	for _, name := range listNames {
+		if err := f(name); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// IsObjNotFoundErr returns true if error means that object is not found. Relevant to Get operations.
+func (b *Bucket) IsObjNotFoundErr(err error) bool {
+	level.Debug(b.logger).Log("msg", "Check if err it's NotFoundError", "err", err)
+	if err == nil {
+		return false
+	}
+	switch errorCode := parseError(err.Error()); errorCode {
+	case "InvalidUri":
+		return true
+	case "BlobNotFound":
+		return true
+	default:
+		return false
+	}
+}
+
+func (b *Bucket) getBlobReader(ctx context.Context, name string, offset, length int64) (io.ReadCloser, error) {
+	level.Debug(b.logger).Log("msg", "Getting blob", "blob", name, "offset", offset, "length", length)
+	if len(name) == 0 {
+		return nil, errors.New("X-Ms-Error-Code: [EmptyContainerName]")
+	}
+	exists, err := b.Exists(ctx, name)
+
+	if !exists {
+		return nil, errors.New("X-Ms-Error-Code: [BlobNotFound]")
+	}
+
+	blobURL := getBlobURL(ctx, b.config.StorageAccountName, b.config.StorageAccountKey, b.config.ContainerName, name)
+
+	props, err := blobURL.GetProperties(ctx, blob.BlobAccessConditions{})
+	if err != nil {
+		level.Error(b.logger).Log("msg", "Cannot get properties for container", "address", name)
+		return nil, err
+	}
+
+	var size int64
+	if length > 0 {
+		size = length
+	} else {
+		size = props.ContentLength() - offset
+	}
+
+	var destBuffer []byte
+
+	destBuffer = make([]byte, size)
+
+	err = blob.DownloadBlobToBuffer(context.Background(), blobURL.BlobURL, offset, length,
+		blob.BlobAccessConditions{}, destBuffer, blob.DownloadFromBlobOptions{
+			BlockSize:   blob.BlobDefaultDownloadBlockSize,
+			Parallelism: uint16(3),
+			Progress:    nil,
+		})
+
+	reader := bytes.NewReader(destBuffer)
+
+	readerCloser := &blobReadCloser{
+		Reader: reader,
+	}
+
+	return readerCloser, err
+}
+
+// Get returns a reader for the given object name.
+func (b *Bucket) Get(ctx context.Context, name string) (io.ReadCloser, error) {
+	return b.getBlobReader(ctx, name, 0, blob.CountToEnd)
+}
+
+// GetRange returns a new range reader for the given object name and range.
+func (b *Bucket) GetRange(ctx context.Context, name string, off, length int64) (io.ReadCloser, error) {
+	return b.getBlobReader(ctx, name, off, length)
+}
+
+// Exists checks if the given object exists.
+func (b *Bucket) Exists(ctx context.Context, name string) (bool, error) {
+	level.Debug(b.logger).Log("msg", "Check if blob exists", "blob", name)
+	b.opsTotal.WithLabelValues(opObjectHead).Inc()
+	blobURL := getBlobURL(ctx, b.config.StorageAccountName, b.config.StorageAccountKey, b.config.ContainerName, name)
+	_, err := blobURL.GetProperties(ctx, blob.BlobAccessConditions{})
+	if b.IsObjNotFoundErr(err) {
```

Comment:
what if it is NOT object not found error because it is any other error (TCP connection closed, timeout etc). I think we need to handle this